scripts/main/upload.js

```diff
@@ -347,209 +445,243 @@ upload.start = {
+									break;
+								case "ImportCancelledException":
+									statusText = lychee.locale["UPLOAD_CANCELLED"];
+									noteText = lychee.locale["UPLOAD_IMPORT_CANCELLED"];
+									break;
+								default:
+									statusText = lychee.locale["UPLOAD_SKIPPED"];
+									noteText = report.message;
+									break;
 							}
-							// Parse the response as JSON.  This will remove
-							// the surrounding '"' characters, unescape any '"'
-							// from the middle, and translate '\n' sequences into
-							// newlines.
-							let jsonResponse;
-							try {
-								jsonResponse = JSON.parse(response);
-							} catch (e) {
-								// Most likely a SyntaxError due to something
-								// that went wrong on the server side.
-								$(lastRowSelector + " .status")
-									.html(lychee.locale["UPLOAD_FAILED"])
-									.addClass("error");
 
+							$(".status", jqEventRow).text(statusText).addClass(severityClass);
+							$(".notice", jqEventRow).text(noteText).show();
+
+							encounteredProblems = true;
+						}
+					}); // forEach (resp)
+					lastReadIdx = reports.length;
+					$(jqRows).scrollTop(topSkip);
+				}; // processIncremental
+
+				/**
+				 * @param {ImportReport[]} reports
+				 */
+				const successHandler = function (reports) {
+					// reports is already JSON-parsed.
+					processIncremental(reports);
+
+					albums.refresh();
+
+					upload.notify(lychee.locale["UPLOAD_IMPORT_COMPLETE"], encounteredProblems ? lychee.locale["UPLOAD_COMPLETE_FAILED"] : null);
+
+					if (albumID === null) lychee.goto();
+					else album.load(albumID);
+
+					if (encounteredProblems) showCloseButton();
+					else basicModal.close();
+				};
+
+				/**
+				 * @this {XMLHttpRequest}
+				 */
+				const progressHandler = function () {
+					/** @type {string} */
+					let response = this.response;
+					/** @type {ImportReport[]} */
+					let reports = [];
+					// We received a possibly partial response.
+					// We must ensure that the last object in the
+					// array is complete and terminate the array.
+					while (response.length > 2 && reports.length === 0) {
+						// Search the last '}', assume that this terminates
+						// the last JSON object, cut the string and terminate
+						// the array with `]`.
+						const fixedResponse = response.substring(0, response.lastIndexOf("}") + 1) + "]";
+						try {
+							// If the assumption is wrong and the last found
+							// '}'  does not terminate the last object, then
+							// `JSON.parse` will fail and tell us where the
+							// problem occurred.
+							reports = JSON.parse(fixedResponse);
+						} catch (e) {
+							if (e instanceof SyntaxError) {
+								const errorPos = e.columnNumber;
+								const lastBrace = response.lastIndexOf("}");
+								const cutResponse = errorPos < lastBrace ? errorPos : lastBrace;
+								response = response.substring(0, cutResponse);
+							} else {
+								// Something else went wrong
 								albums.refresh();
```

Comment:
No, not by intention.

Reply:
I must revise my previous statement. The error case above simply did not exist in the previous code, because the previous code used a home-brewed newline-terminated message format, nothing which was proper JSON.

Here we handle the case, that `JSON.parse` fails due to some other reason than a syntax error. A syntax error may occur due to a yet incomplete response and is handled properly, by stripping off the tail of the incomplete response down to the nearest closing `}`.

On current master, the closest thing which might match is the following code

https://github.com/LycheeOrg/Lychee-front/blob/master/scripts/main/upload.js#L404-L408

So the current code does not handle a comparable situation at all, but the code were simply caught in an endless loop.

With respect to the current code I would argue that the else-branch cited above should never be called. According to https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON/parse#exceptions `JSON.parse` only throws `SyntaxError` and nothing else.

The else-branch above only exist, because JS is not type-safe and I needed something there. If JS would support it, this would be the classical case for an `assert`.

Reply:
I was referring specifically to:

https://github.com/LycheeOrg/Lychee-front/blob/feba79086476c7c522052b9b8d4c81b397a888c3/scripts/main/upload.js#L549-L551

But now that I look at the new code again, I can see that it's quite different so no simple comparisons should be made.

In practical sense, I believe the old code on the master is there to handle the server crashes (which will likely result in an unfinished, and thus unparsable, JSON string). I know that error handling is now very different (as in, much better in general), but does that apply to streaming responses like here as well? I.e., what do you expect to happen if the server crashes during import? Ideally we would of course prefer that it never does, but... It would be interesting to introduce a deliberate bug on the server side during testing to see if the front end reacts as expected -- have you tried that? (adding it to my todo list, just in case...)

Reply:
> In practical sense, I believe the old code on the master is there to handle the server crashes (which will likely result in an unfinished, and thus unparsable, JSON string).

This case should lead to a `SynraxError` being thrown and thus is handled. The `else`-branch here should actually only be called, if the JS engine throws an exception which it should not acc. to the JS specification. So maybe, we should bail out with a `window.message` and log a critical emergency to the console.

> I.e., what do you expect to happen if the server crashes during import?

We have to distinguish two cases here:

 1. the backend encounters a bug and throws an exception
 2. the backend encounters a hard crash, e.g. the webserver or PHP process is abruptly killed by the server

In the first case, the exception should be inserted as a `EventReport` in the streamed JSON response. The frontend should display that error.

In the second case, the streamed response simply stops and no data more data is coming in. In this case, I expect the frontend to continue to display the import dialog forever until it is cancelled by the user.

> It would be interesting to introduce a deliberate bug on the server side during testing to see if the front end reacts as expected -- have you tried that? (adding it to my todo list, just in case...)

I did not test that intentionally. I tested case 1 accidentally several times during coding due to real bugs. But it is still worth to be tested again. I never tested case 2.